lusidtools/cocoon/cocoon.py

```diff
@@ -901,11 +904,342 @@ async def _construct_batches(
+        for code, effective_at in zip(sync_batch["codes"], sync_batch["effective_at"],):
+            # Append a tuple of (code, effective_at) to the code_tuples list
+            code_tuples.append((code, effective_at))
+    return list(set(code_tuples))
+
+
+def return_unmatched_transactions(
+    api_factory: lusid.utilities.ApiClientFactory, scope: str, code: str
+):
+    """
+    Call the get transactions api and only return those transactions with unresolved identifiers.
+
+    Multiple pages of responses for the get_transactions api call are handled.
+
+    Parameters
+    ----------
+    api_factory : lusid.utilities.ApiClientFactory api_factory
+        The api factory to use
+    scope : str
+        The scope of the resource to load the data into
+    code : str
+        The code of the portfolio containing the transactions we want to return
+
+    Returns
+    -------
+    A list of objects with the structure { transaction_id: instrument_identifiers }.
+    """
+    transactions_api = api_factory.build(lusid.api.TransactionPortfoliosApi)
+    done = False
+    next_page = None
+    unmatched_transactions = []
+
+    # We need to handle the possibility of paginated results
+    while not done:
+
+        # There must be a filter included so only transactions with unmatched instruments are returned
+        kwargs = {
+            "scope": scope,
+            "code": code,
+            "filter": "instrumentUid eq'LUID_ZZZZZZZZ'",
+        }
+
+        if next_page is not None:
+            kwargs["page"] = next_page
+
+        response = transactions_api.get_transactions(**kwargs)
+
+        unmatched_transactions.extend(
+            [
+                {response.transaction_id: response.instrument_identifiers}
+                for response in response.values
+            ]
+        )
+
+        next_page = response.next_page
+        done = response.next_page is None
+
+    return unmatched_transactions
+
+
+def filter_unmatched_transactions(
+    data_frame: pd.DataFrame, mapping_required: dict, unmatched_transactions: list,
+):
+    """
+    This method will take the full list of unmatched transactions and remove any transactions that were not
+    part of the current upload with load_from_data_frame.
+
+    Parameters
+    ----------
+    data_frame : pd.DataFrame
+        The DataFrame containing the data
+    mapping_required : dict
+        The required mapping from load_from_data_frame that helps identify the transaction_id column in the DataFrame
+    unmatched_transactions : list
+         A list of objects with the structure { transaction_id: instrument_identifiers }
+
+    Returns
+    -------
+    A filtered list of unmatched_transactions.
+    """
+    # Create a list var to hold the filtered transactions to be returned
+    filtered_unmatched_transactions = []
+
+    # Create a unique list of transaction_ids from the dataframe
+    valid_txids = data_frame[mapping_required.get("transaction_id")].unique()
```

Comment:
make this a set so the `in` check is O(1) lookup

Reply:
for dict.get(): Average case O(1); Amortised analysis case O(N)

(https://www.geeksforgeeks.org/complexity-cheat-sheet-for-python-operations/)